libazureinit/src/config.rs

```diff
@@ -0,0 +1,681 @@
+/// directory structure.
+///
+/// # Arguments
+///
+/// * `path` - Optional path to a configuration file or directory.
+impl Config {
+    pub fn load(path: Option<PathBuf>) -> Result<Config, Error> {
+        let mut config = Config::default();
+
+        if let Some(cli_config) = path {
+            if cli_config.is_dir() {
+                config = Self::load_from_directory(cli_config)?;
+            } else {
+                config = Self::load_from_file(cli_config)?;
+            }
+        }
+
+        Ok(config)
+    }
+
+    /// Loads the configuration from a single file.
+    ///
+    /// # Arguments
+    ///
+    /// * `file_path` - Path to the configuration file.
+    fn load_from_file(file_path: PathBuf) -> Result<Config, Error> {
+        let content = fs::read_to_string(file_path)?;
+        toml::from_str::<Config>(&content).map_err(|e| {
+            tracing::error!("Failed to parse configuration file: {:?}", e);
+            Error::Io(std::io::Error::new(
+                std::io::ErrorKind::InvalidData,
+                format!("Failed to parse TOML config file: {:?}", e),
+            ))
+        })
+    }
+
+    /// Loads the configuration from a directory.
+    ///
+    /// If the directory contains a `azure-init.toml` file, it loads that file first. It then
+    /// loads all `.toml` files from a subdirectory named `azure-init.d`, merging them with the
+    /// base configuration in alphabetical order.
+    ///
+    /// # Arguments
+    ///
+    /// * `dir` - Path to the configuration directory
+    fn load_from_directory(dir: PathBuf) -> Result<Config, Error> {
+        let mut config = Config::default();
+
+        let base_config_path = dir.join("azure-init.toml");
+        if base_config_path.exists() {
+            config = config.merge(Self::load_from_file(base_config_path)?);
+        }
+
+        let d_dir = dir.join("azure-init.d");
+        if d_dir.is_dir() {
+            let mut toml_files: Vec<_> = fs::read_dir(d_dir)?
+                .filter_map(|entry| {
+                    let entry = entry.ok()?;
+                    let path = entry.path();
+                    if path.extension()?.to_str()? == "toml" {
+                        Some(path)
+                    } else {
+                        None
+                    }
+                })
+                .collect();
+
+            toml_files.sort();
+
+            for file_path in toml_files {
+                config = config.merge(Self::load_from_file(file_path)?);
+            }
+        }
+
+        Ok(config)
+    }
+
+    /// Merges two configurations, giving priority to values from `other`.
+    ///
+    /// This method combines two configurations, with each field in `other` overwriting the
+    /// corresponding field in `self`. This allows for merging multiple configurations in
+    /// order of precedence, for example, applying CLI-specified configurations over defaults.
+    fn merge(mut self, other: Config) -> Config {
```

Comment:
Oh, I see what you're saying here. I could see there being value in making merge happen at the individual key level! I think the change would just be merging individual keys within each section rather than one large merge the way it currently is. 


Something like this would then be in the struct for each field: 
```rust
pub struct Ssh {
    pub key_a: Option<String>,
    pub key_b: Option<String>,
    pub key_c: Option<String>,
}

impl Ssh {
    pub fn merge(mut self, other: Ssh) -> Ssh {
        if let Some(key_a) = other.key_a {
            self.key_a = Some(key_a);
        }
        if let Some(key_b) = other.key_b {
            self.key_b = Some(key_b);
        }
        if let Some(key_c) = other.key_c {
            self.key_c = Some(key_c);
        }
        self
    }
}
```
I don't think that would be a tough change, if it's one that would make more sense for the user. 
@cjp256 and @anhvoms, any thoughts here on section level merge vs. individual key merge?